lib/test.js

```diff
@@ -53,6 +53,77 @@ function plan(count) {
 	this.plan(count, captureStack(this.plan));
 }
 
+function tryTest(fn) {
+	const args = [];
+	for (let i = 1, len = arguments.length; i < len; i++) {
+		args.push(arguments[i]);
+	}
+
+	let commitCalled = false;
+	let discarded = false;
+	let finished = false;
+
+	const commitFn = (ret, toCommit) => {
+		if (commitCalled) {
+			if (discarded) {
+				this.saveFirstError(new Error('The discard() was already called'));
+			} else {
+				this.saveFirstError(new Error('The commit() was already called'));
+			}
+			return;
+		}
+		commitCalled = true;
+
+		if (ret.passed) {
+			this.countFinishedAttemptAssertion(toCommit);
+		} else {
+			this.countFailedAttemptAssertion(ret.error, toCommit);
+		}
+
+		if (toCommit) {
+			for (const log of ret.logs) {
+				this.addLog(log);
+			}
+		}
```

Comment:
It might be better to have one `commitAttempt(ret)` function call which performs all the relevant actions based on `ret`'s state.

Reply:
Could you elaborate?

Reply:
The method would sit on `Test.prototype`, so you can call `this.commitAttempt(ret)`. That implementation can safely access Test internals and you wouldn't need the additional count methods etc.